File: WeatherCell.java
import java.util.Random;

public class WeatherCell {
    public double temperature;  // in Kelvin
    public double pressure;     // in Pascals
    public double humidity;     // as a percentage (0 to 100)

    // Physical constants
    public static final double AIR_DENSITY = 1.225;             // kg/m^3 at sea level
    public static final double SPECIFIC_HEAT_CAPACITY = 1005;   // J/(kg·K) for air

    // Random values for initialization
    public WeatherCell() {
        Random rand = new Random();
        // Random temperature between 0°C and 30°C (in Kelvin)
        this.temperature = 273.15 + rand.nextDouble() * 30.0;
        // Random pressure around 100 kPa +/- 2500 Pa
        this.pressure = 100000 + rand.nextDouble() * 5000 - 2500;
        // Random humidity between 0% and 100%
        this.humidity = rand.nextDouble() * 100.0;
    }

    // Constructor for specific values
    public WeatherCell(double temperature, double pressure, double humidity) {
        this.temperature = temperature;
        this.pressure = pressure;
        this.humidity = humidity;
    }

    // Update cell state based on simplified physics
    public WeatherCell calculateNewState(WeatherCell[] neighbors) {
        double tempChange = 0;
        double pressureChange = 0;
        double humidityChange = 0;
        int count = 0;

        for (WeatherCell neighbor : neighbors) {
            if (neighbor != null) {
                // Temperature exchange based on heat transfer
                tempChange += (neighbor.temperature - this.temperature);
                // Pressure change based on fluid dynamics (simplified)
                pressureChange += (neighbor.pressure - this.pressure);
                // Humidity diffusion
                humidityChange += (neighbor.humidity - this.humidity);
                count++;
            }
        }

        if (count > 0) {
            // Simple averaging for demonstration purposes
            tempChange /= count;
            pressureChange /= count;
            humidityChange /= count;
        }

        // Update temperature using a simplified heat equation
        double newTemperature = this.temperature + (tempChange * 0.1); // Coefficient represents thermal conductivity

        // Update pressure based on simplified fluid dynamics
        double newPressure = this.pressure + (pressureChange * 0.1); // Coefficient represents pressure diffusion

        // Update humidity
        double newHumidity = this.humidity + (humidityChange * 0.1); // Coefficient represents humidity diffusion

        // Ensure values stay within realistic bounds
        newTemperature = Math.max(200.0, Math.min(newTemperature, 330.0));     // Limit temperature between -73°C and 57°C
        newPressure = Math.max(80000.0, Math.min(newPressure, 120000.0));      // Limit pressure between 80 kPa and 120 kPa
        newHumidity = Math.max(0.0, Math.min(newHumidity, 100.0));             // Limit humidity between 0% and 100%

        return new WeatherCell(newTemperature, newPressure, newHumidity);
    }
}

File: WeatherCell.c
#include "WeatherCell.h"
#include <stdlib.h>
#include <time.h>
#include <math.h>

// Helper function to get random value in range
static double random_range(double min, double max) {
    double scale = rand() / (double) RAND_MAX;
    return min + scale * (max - min);
}

// Default constructor with random initialization
WeatherCell* WeatherCell_new(void) {
    WeatherCell* cell = (WeatherCell*)malloc(sizeof(WeatherCell));
    if (cell == NULL) return NULL;
    
    // Initialize with random but realistic values
    cell->temperature = random_range(273.15, 313.15);  // 0°C to 40°C
    // Calculate pressure from temperature
    cell->pressure = cell->temperature * MOLES * GAS_CONSTANT;
    cell->rHumidity = random_range(0, 100);             // 0-100%
    // Specific humidity is calculated based off relative humidity
    double celsius = cell->temperature - 273.15;
    cell->sHumidity = ((cell->rHumidity/100) * 0.61121 * exp((18.678-(celsius/234.5))*(celsius/(257.14 + celsius)))) / (461.5 * cell->temperature * 100) * 1000; // in kg/m^3
    //Calculate humid heat capacity
    cell->hCapacity = SPECIFIC_HEAT_CAPACITY + (1.82 * (cell->sHumidity/(AIR_DENSITY - cell->sHumidity)));
    
    return cell;
}

// Constructor with specific values
WeatherCell* WeatherCell_new_with_values(double temperature, double pressure, double humidity) {
    WeatherCell* cell = (WeatherCell*)malloc(sizeof(WeatherCell));
    if (cell == NULL) return NULL;
    
    cell->temperature = temperature;
    double celsius = temperature - 273.15;
    cell->pressure = pressure;
    cell->rHumidity = humidity > 100 ? 100 : (humidity < 0 ? 0 : humidity);
    cell->sHumidity = ((cell->rHumidity/100) * 0.61121 * exp((18.678-(celsius/234.5))*(celsius/(257.14 + celsius)))) / (461.5 * cell->temperature * 100) * 1000; // in kg/m^3
    cell->hCapacity = SPECIFIC_HEAT_CAPACITY + (1.82 * (cell->sHumidity/(AIR_DENSITY - cell->sHumidity)));
    
    return cell;
}

// Calculate new state based on neighbors
WeatherCell* WeatherCell_calculate_new_state(WeatherCell* cell, WeatherCell* neighbors[4]) {
    if (cell == NULL) return NULL;
    
    // Create new cell for next state
    WeatherCell* new_cell = WeatherCell_new_with_values(
        cell->temperature,
        cell->pressure,
        cell->rHumidity
    );
    if (new_cell == NULL) return NULL;

    // Count valid neighbors
    int valid_neighbors = 0;
    double heatChange = 0;
    double tempChange = 0;
    double pressureChange = 0;
    double sHumidityChange = 0;
    double newTemperature = cell->temperature;
    double newPressure = cell->pressure;
    double newRHumidity = cell->rHumidity;

    // Calculate averages including current cell
    valid_neighbors = 0;  // Count self

    // Add neighbor values
    for (int i = 0; i < 4; i++) {
        if (neighbors[i] != NULL) {
            // Temperature exchange based on heat transfer
                double htc = ((neighbors[i]->temperature - cell->temperature)/1.444) + 10;
                heatChange += htc * (neighbors[i]->temperature - cell->temperature); // watt/m^2
                // Pressure change based on fluid dynamics (simplified)
                pressureChange += neighbors[i]->pressure;
                // Simplified Humidity diffusion
                sHumidityChange += neighbors[i]->sHumidity;
                valid_neighbors++;
        }
    }

    // Calculate final averages
    if (valid_neighbors > 0) {
        tempChange = heatChange/(AIR_DENSITY * cell->hCapacity * 1000);             //simplified temperature exchange through convection
        newTemperature = cell->temperature + tempChange;
        double celsius = newTemperature - 273.15;
        newPressure = (cell->pressure + pressureChange)/(valid_neighbors + 1);       //simplified pressure exchange
        double tempsHumidity = (cell->sHumidity + sHumidityChange)/(valid_neighbors + 1);    //VERY simplified humidity exchange
        newRHumidity = (tempsHumidity * 461.5 * newTemperature * 100)/(0.61121 * exp((18.678-(celsius/234.5))*(celsius/(257.14 + celsius))))/10;          
    }

    
    // Update new cell state with smoothed values and adjustments
    new_cell->temperature = newTemperature > 300 ? 300 : (newTemperature < 200 ? 200 : newTemperature);
    new_cell->pressure = newPressure > 120000 ? 120000 : (newPressure < 80000 ? 80000 : newPressure);
    new_cell->rHumidity = newRHumidity > 100 ? 100 : (newRHumidity < 0 ? 0 : newRHumidity);

    return new_cell;
}

// Memory management
void WeatherCell_destroy(WeatherCell* cell) {
    free(cell);
}
